### ISSUE TITLE
Write unit tests and setup code coverage

### ISSUE THREAD

**Issue by wizard-28** [pacstall/pacstall#385]
## What are unit tests? Why does it matter?

https://realpython.com/python-testing/

TL;DR: Basically writing separate code to test individual elements (functions etc) of the main code base. The advantage here is that after defining the tests the whole testing process becomes easier and automated. Though I suggest you read through that article I linked.

## What is code coverage?

Code coverage is a metric used to describe the degree to which the source code of a program is tested. A program with high code coverage has been more thoroughly tested and has a lower chance of containing software bugs than a program with low code coverage.

Codacy has a Code Coverage reporter. Which we could use to display out Code Coverage in the README, if we wanted.

## Which unit testing framework and code coverage measuring tool should we use?

[PyTest](https://docs.pytest.org/en/6.2.x/) seems to be the most popular unit testing framework for Python. Source code [here](https://github.com/pytest-dev/pytest/).

Codacy supports reports from [Coverage.py](https://coverage.readthedocs.io/en/coverage-5.0.3/) for Python, so I think we should use that to generate Coverage reports. Source code [here](https://github.com/nedbat/coveragepy/tree/master)

## Additional Info

[Codacy guide](https://docs.codacy.com/coverage-reporter/) on setting up Code Coverage
[Codacy blog](https://blog.codacy.com/a-guide-to-code-coverage-part-1-code-coverage-explained/) on setting up Code Coverage

[Alternative ways of running Coverage Reporter](https://docs.codacy.com/coverage-reporter/alternative-ways-of-running-coverage-reporter/) might be useful if we used the GitHub Action part.

I think it'll be best if we write a Code Coverage reporter action from scratch. That runs the tests, generates Code Coverage report, and then uploads it to Codacy.
